util/gcs/read_test.go

```diff
@@ -212,6 +222,378 @@ func TestReadJSON(t *testing.T) {
 	}
 }
 
+type fakeIterator struct {
+	objects []storage.ObjectAttrs
+	idx     int
+	err     int
+}
+
+func (fi *fakeIterator) Next() (*storage.ObjectAttrs, error) {
+	if fi.idx > len(fi.objects) {
+		return nil, iterator.Done
+	}
+	if fi.idx == fi.err {
+		return nil, errors.New("injected fakeIterator error")
+	}
+
+	o := fi.objects[fi.idx]
+	fi.idx++
+	return &o, nil
+}
+
+type fakeObject struct {
+	data    string
+	listing []storage.ObjectAttrs
+	open    error
+	read    error
+}
+
+func subdir(prefix string) storage.ObjectAttrs {
+	return storage.ObjectAttrs{Prefix: prefix}
+}
+
+func link(name, other string) storage.ObjectAttrs {
+	return storage.ObjectAttrs{
+		Metadata: map[string]string{"x-goog-meta-link": other},
+		Name:     name,
+	}
+}
+
+type fakeInterrogator map[Path]fakeObject
+
+func (fi fakeInterrogator) Objects(_ context.Context, path Path, _ string) Iterator {
+	return &fakeIterator{objects: fi[path].listing}
+}
+
+func (fi fakeInterrogator) Open(ctx context.Context, path Path) (io.ReadCloser, error) {
+	o, ok := fi[path]
+	if !ok {
+		return nil, fmt.Errorf("wrap not exist: %w", storage.ErrObjectNotExist)
+	}
+	if o.open != nil {
+		return nil, o.open
+	}
+	return ioutil.NopCloser(&fakeReader{
+		buf:     bytes.NewBufferString(o.data),
+		readErr: o.read,
+	}), nil
+}
+
 // TODO(fejta): TestStarted
+func TestStarted(t *testing.T) {
```

Comment:
Do we need the boilerplate for TestStarted? (Or was this in-progress?)

Reply:
I have a PR out to fill this out, just felt like this PR was already large enough